### DISCUSSION TITLE
匹配应答怎么只保留IPV4结果，忽略IPV6

### DISCUSSION THREAD

**Discussion by muziling** [IrineSistiana/mosdns#106] [Q&A]
匹配应答怎么只保留IPV4结果，忽略IPV6

**Comment by streetwind** [comment]
相同问题，没有找到怎么配置方案

**Comment by laggardkernel** [comment]
你是说一个DNS响应中同时含有IPv4和IPv6数据，想要移除IPv4数据？这种功能好像不支持，涉及修改响应数据包。我只听说 Unbound + Python 扩展可以实现。

**Reply by streetwind** [reply]
嗯，大概是这意思，我是希望双栈域名，有IPV4的情况忽略IPV6，因为现在系统都是优先ipv6连接，没有IPV4返回IPV6使用IPV6连接。虽然可以修改系统浏览器之类的实现这样的效果主要太麻烦了

**Reply by IrineSistiana** [reply]
一个DNS响应中好像不会同时含有IPv4和IPv6数据吧。
解析双栈域域名应该是客户端发了两个请求，一个 A,一个 AAAA。一般情况把 AAAA 请求屏蔽就可以了。

**Reply by streetwind** [reply]
nslookup aliyun.com 114.114.114.114
太会返回多个IPV4和IPV6
dns请求的时候可以限制A或者AAAA 可是我无法知道哪些域名是纯IPV6域名。我全部屏蔽AAAA请求类型会导致纯IPV6网站无法访问

**Reply by IrineSistiana** [reply]
目前实现不了。这个特性我会安排上，很实用。

**Reply by streetwind** [reply]
这特性应该是针对返回请求修改，现在ipv6和ipv4同时存在的复杂网络，系统默认又已IPV6优先。
IPV6的不稳定性和高延时，所有才特意希望有这样的功能来过滤应答请求，优先使用IPV4去除IPV6的DNS响应不用每台系统修改优先使用IPV4
    感谢作者能对这特性开发

**Reply by IrineSistiana** [reply]
master 已经有这个功能了，可以自行编译体验下。

wiki 说明：

---
### _prefer_ipv4/6 智能过滤 ipv4/6

`_prefer_ipv4` 和 `_prefer_ipv6` 是预初始化插件，可直接使用。

下面为了表述方便，仅以 `_prefer_ipv4` 为例。 `_prefer_ipv6` 同理。

`_prefer_ipv4` 处理 AAAA 请求时会格外产生一个新 A 请求。如果格外请求成功获得域名的 A 记录，则源请求会被屏蔽。因此 `_prefer_ipv4` 会在域名存在 A 地址的前提下屏蔽 AAAA 请求。相比强行屏蔽所有 AAAA 请求，`_prefer_ipv4` 不会影响纯 IPv6 域名的解析。

注意事项:

- 由于 `_prefer_ipv4` 会生成新请求，理论上会使上游流量增加 50%。(支持双栈的客户端会同时发送 A 和 AAAA 两个请求，但插件在处理 AAAA 时会格外产生一个 A 请求，共三个请求被发送至上游)。如果在意性能，可在 `_prefer_ipv4` 紧接 `_single_flight` 去除重复的 A 请求。

**Reply by streetwind** [reply]
已编译，试用了。能达到目标需求。
我不知道作者你的设计思路，如果在请求处理在我的想法是，如果设置是优先IPV4，请求时A地址有返回就不请求AAAA类型。
另外就是在返回处理，因为dns请求默认会请求A和AAAA类型，筛选A类型优先

**Reply by IrineSistiana** [reply]
没么简单。A 和 AAAA 是两个**完全独立的请求**。即使客户端会同时发这两个请求 mosdns 也不能把它们关联。（实际上大部分情况根本不是一起发，取决于客户端的缓存A/AAAA谁先过期。）

**Reply by streetwind** [reply]
嗯，可能我想得太简单。没有仔细看过mosdns代码里面的执行逻辑，那作者你考虑在mosdns的缓存机制上做文章那的方案可行不。

**Reply by IrineSistiana** [reply]
> 在mosdns的缓存机制上做文章那的方案可行不

不是很明白....

**Reply by streetwind** [reply]
mosdns 有自己的缓存机制，如果查询域名请求缓存没有会发送A/AAAA请求，这两个请求都保存到缓存里面。
如果配置A类型优先，取缓存时不取AAAA结果响应下游客户端。我不清楚mosdns的缓存机制所以只是这样的想法
 不过谢谢作者能开发此功能，对我个人而言这个功能很实用

**Reply by IrineSistiana** [reply]
把 cache 插件放在 prefer_ip 后面就行了。这样 cache 就能缓存prefer_ip发出的两种请求。如果有缓存会直接返给 prefer_ip。

~mosdns 的运行框架就是这样设计的~

**Reply by hplee0120** [reply]
> master 已经有这个功能了，可以自行编译体验下。
> 
> wiki 说明：
> 
> ### _prefer_ipv4/6 智能过滤 ipv4/6
> `_prefer_ipv4` 和 `_prefer_ipv6` 是预初始化插件，可直接使用。
> 
> 下面为了表述方便，仅以 `_prefer_ipv4` 为例。 `_prefer_ipv6` 同理。
> 
> `_prefer_ipv4` 处理 AAAA 请求时会格外产生一个新 A 请求。如果格外请求成功获得域名的 A 记录，则源请求会被屏蔽。因此 `_prefer_ipv4` 会在域名存在 A 地址的前提下屏蔽 AAAA 请求。相比强行屏蔽所有 AAAA 请求，`_prefer_ipv4` 不会影响纯 IPv6 域名的解析。
> 
> 注意事项:
> 
> * 由于 `_prefer_ipv4` 会生成新请求，理论上会使上游流量增加 50%。(支持双栈的客户端会同时发送 A 和 AAAA 两个请求，但插件在处理 AAAA 时会格外产生一个 A 请求，共三个请求被发送至上游)。如果在意性能，可在 `_prefer_ipv4` 紧接 `_single_flight` 去除重复的 A 请求。

请问这个_prefer_ipv4参数是不是意味着无论返回的ipv4 dns是否能连通都会过滤掉AAAA然后使用这个ipv4的dns。某些域名 比如这个  http://hdtv.haust.edu.cn/ 。在v4环境下是无法打开的。使用_prefer_ipv4这个参数后同样无法打开。而使用之前的保留国内双栈，过滤掉国外的AAAA的方案下是能正常打开的。这是正常现象吗

**Reply by streetwind** [reply]
那你应该优先使用ipv6过滤掉ipv4，你使用_prefer_ipv4参数后如果dns返回有ipv4会过滤掉ipv6地址

**Reply by hplee0120** [reply]
> 那你应该优先使用ipv6过滤掉ipv4，你使用_prefer_ipv4参数后如果dns返回有ipv4会过滤掉ipv6地址

优先使用ipv6丢弃ipv4。按照目前的ipv6环境感觉不太合理。这个参数的主要适用环境应该是优先使用ipv4，兼顾纯ipv6网站？但是比如这种返回双栈记录，但是只有ipv6能连通的应该不太好控制。 我暂时还是使用之前的老方案 保留国内双栈 过滤国外的ipv6算了。

**Reply by streetwind** [reply]
你可以在配置的时候特定网站优先使用ipv6 其他一样使用ipv4优先，像这种返回的ipv4无法联通是不好控制。你没办法叫一个dns服务去检测返回ip的联通性啊

**Comment by zjzxj** [comment]
如果上游是tcp协议，_prefer_ipv4似乎无效

**Comment by ilafz** [comment]
openwrt的dnsmasq里面把禁止解析 IPv6 DNS 记录勾上就行
